File: lib/rust/src/benchmark.rs
use std::{
    env,
    fmt::Display,
    fs::File,
    io::{self, BufRead, BufReader},
    num::ParseIntError,
    path::Path,
    time::{Duration, Instant},
};

const NANOS_PER_MILLI: f64 = 1_000_000f64;
const INITIAL_RESULT_CAPACITY: usize = 1_000;

#[derive(Debug, Default)]
pub struct BenchmarkArguments {
    pub run_ms: u64,
    pub warmup_ms: u64,
    pub program_args: Vec<String>,
}

#[derive(Debug, Default)]
pub struct TimedResult<T> {
    #[allow(dead_code)]
    pub elapsed_total: Duration,
    pub elapsed: Duration,
    pub value: T,
}

#[derive(Debug, Default)]
pub struct BenchmarkResult<T> {
    pub mean: Duration,
    pub std_dev: Duration,
    pub min: Duration,
    pub max: Duration,
    pub results: Vec<TimedResult<T>>,
}

impl<T> BenchmarkResult<T> {
    pub fn runs(&self) -> usize {
        self.results.len()
    }

    pub fn last_result(&self) -> Option<&T> {
        Some(&self.results.last()?.value)
    }

    pub fn format_output(&self, check_value: impl Display) -> String {
        format!(
            "{:.6},{:.6},{:.6},{:.6},{},{}",
            (self.mean.as_nanos() as f64 / NANOS_PER_MILLI),
            (self.std_dev.as_nanos() as f64 / NANOS_PER_MILLI),
            (self.min.as_nanos() as f64 / NANOS_PER_MILLI),
            (self.max.as_nanos() as f64 / NANOS_PER_MILLI),
            self.runs(),
            check_value
        )
    }

    pub fn print_output(&self, check_value: impl Display) {
        let formatted_output = self.format_output(check_value);
        println!("{}\n", formatted_output);
    }
}

pub struct BenchmarkContext<T> {
    benchmark_fn: Box<dyn FnMut() -> T + 'static>,
}

/// Type T must be the return value of the benchmarked function.
impl<T> BenchmarkContext<T> {
    /// Create a new benchmark context with a closure to benchmark.
    pub fn new(benchmark_fn: impl FnMut() -> T + 'static) -> Self {
        Self {
            benchmark_fn: Box::new(benchmark_fn),
        }
    }

    /// run_ms:
    ///   0: don't run
    ///   1: check-output run
    pub fn run(&mut self, run_ms: u64) -> Option<BenchmarkResult<T>> {
        let mut benchmark_results: Vec<TimedResult<T>> =
            Vec::with_capacity(INITIAL_RESULT_CAPACITY);

        if run_ms == 0 {
            return None;
        } else if run_ms > 1 {
            // start with a status dot, but not if this is a check-output run
            eprint!(".");
        }

        let run_duration = Duration::from_millis(run_ms);

        let mut elapsed_total = Duration::ZERO;
        let mut last_time = Instant::now();

        while elapsed_total < run_duration {
            let time_start = Instant::now();

            let current_value: T = (self.benchmark_fn)();

            let time_end = Instant::now();

            // print status dots every second if it isn't a check-output run
            if run_ms > 1 && time_start.duration_since(last_time) > Duration::from_secs(1) {
                last_time = time_end;
                eprint!(".");
            }

            let elapsed = time_end.duration_since(time_start);
            elapsed_total = elapsed_total.saturating_add(elapsed);

            benchmark_results.push(TimedResult::<T> {
                elapsed_total,
                elapsed,
                value: current_value,
            });
        }

        // add newline for non check-output runs
        if run_ms > 1 {
            eprintln!();
        }

        // calculate timings after completing the run
        // (no unnecessary work during benchmark loop)
        let min: Duration = self.min(&benchmark_results)?;
        let max: Duration = self.max(&benchmark_results)?;
        let mean: Duration = self.mean(&benchmark_results)?;
        let std_dev: Duration = self.standard_deviation(&benchmark_results, mean)?;

        Some(BenchmarkResult::<T> {
            mean,
            std_dev,
            min,
            max,
            results: benchmark_results,
        })
    }

    pub fn benchmark(&mut self, run_ms: u64, warmup_ms: u64) -> Option<BenchmarkResult<T>> {
        // perform warmup runs
        self.run(warmup_ms);

        // perform benchmark runs
        self.run(run_ms)
    }

    fn min(&mut self, results: &[TimedResult<T>]) -> Option<Duration> {
        results.iter().map(|result| result.elapsed).min()
    }

    fn max(&mut self, results: &[TimedResult<T>]) -> Option<Duration> {
        results.iter().map(|result| result.elapsed).max()
    }

    fn mean(&mut self, results: &[TimedResult<T>]) -> Option<Duration> {
        if results.is_empty() {
            return None;
        }

        // needed as u32 for checked_div
        let results_length = u32::try_from(results.len()).ok()?;

        results
            .iter()
            .map(|result| result.elapsed)
            .sum::<Duration>()
            .checked_div(results_length)
    }

    fn standard_deviation(
        &mut self,
        results: &[TimedResult<T>],
        mean: Duration,
    ) -> Option<Duration> {
        if results.is_empty() {
            return None;
        }

        let diffs = results.iter().map(|result| result.elapsed.abs_diff(mean));
        let sum_squared_diffs = diffs
            .map(|diff| diff.as_nanos() * diff.as_nanos())
            .sum::<u128>();
        let std_dev = ((sum_squared_diffs as f64) / (results.len() as f64)).sqrt();

        Some(Duration::from_nanos(std_dev as u64))
    }
}

// Helper Functions

pub fn load_arguments() -> Result<BenchmarkArguments, ParseIntError> {
    let args_cli: Vec<String> = env::args().collect();

    if args_cli.len() < 3 {
        eprintln!(
            "Usage: {} <run_ms> <warmup_ms> ...<program args>",
            args_cli[0]
        );
        std::process::exit(1);
    }

    let (bench_args, program_args) = args_cli.split_at(3);

    let run_ms: u64 = bench_args[1].parse()?;
    let warmup_ms: u64 = bench_args[2].parse()?;

    Ok(BenchmarkArguments {
        run_ms,
        warmup_ms,
        program_args: program_args.iter().map(String::from).collect(),
    })
}

#[allow(dead_code)]
pub fn file_read_lines(file_path: impl AsRef<Path>) -> io::Result<Vec<String>> {
    let file = File::open(file_path)?;
    let buffered_reader = BufReader::new(file);

    buffered_reader.lines().collect()
}

File: lib/zig/benchmark.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Timer = std.time.Timer;

const INITIAL_CAPACITY: usize = 1_000;

pub const ArgumentList = struct {
    run_ms: usize,
    warmup_ms: usize,
    program_args: [][]u8,
};

pub fn FnReturnType(comptime function: anytype) type {
    return @typeInfo(@TypeOf(function)).Fn.return_type.?;
}

pub fn TimedResult(comptime T: type) type {
    return struct {
        elapsed_total: u64,
        elapsed: u64,
        value: T,
    };
}

pub fn BenchmarkResult(comptime T: type) type {
    return struct {
        const Self = @This();

        mean_ms: f64,
        std_dev_ms: f64,
        min_ms: f64,
        max_ms: f64,
        runs: usize,
        results: std.ArrayList(TimedResult(T)),

        pub fn lastResult(self: *const Self) T {
            return self.results.getLast().value;
        }

        pub fn formatOutput(self: *const Self, allocator: Allocator, check_value: anytype) ![]u8 {
            return std.fmt.allocPrint(
                allocator,
                "{d:.6},{d:.6},{d:.6},{d:.6},{d},{any}",
                .{
                    self.mean_ms,
                    self.std_dev_ms,
                    self.min_ms,
                    self.max_ms,
                    self.runs,
                    check_value,
                },
            );
        }

        pub fn printOutput(self: *const Self, allocator: Allocator, check_value: anytype) !void {
            const formatted_output = try self.formatOutput(allocator, check_value);

            const stdout = std.io.getStdOut().writer();
            try stdout.print("{s}\n", .{formatted_output});
        }
    };
}

pub fn createContext(comptime benchmark_fn: anytype) type {
    // store the return type for the working function
    const return_type = FnReturnType(benchmark_fn);

    return struct {
        const Self = @This();

        /// run_ms:
        ///   0: don't run
        ///   1: check-output run
        /// args:
        ///   e.g.: .{ allocator, &arg1, arg2 }
        pub fn run(allocator: Allocator, run_ms: usize, args: anytype) !?BenchmarkResult(return_type) {
            const stderr = std.io.getStdErr().writer();

            var benchmark_results = try std.ArrayList(TimedResult(return_type)).initCapacity(allocator, INITIAL_CAPACITY);

            if (run_ms == 0) {
                return null;
            } else if (run_ms > 1) {
                // start with a status dot, but not if this is a check-output run
                try stderr.writeAll(".");
            }

            const run_ns: usize = run_ms * 1_000_000;

            var elapsed_total: u64 = 0;

            var timer = try Timer.start();
            var last_time = timer.read();

            while (elapsed_total < run_ns) {
                const time_start = timer.read();

                const current_value = @call(.auto, benchmark_fn, args);

                const time_end = timer.read();

                // print status dots every second if it isn't a check-output run
                if (run_ms > 1 and (time_start - last_time) > 1_000_000_000) {
                    last_time = time_end;
                    try stderr.writeAll(".");
                }

                const elapsed = time_end - time_start;
                elapsed_total += elapsed;

                try benchmark_results.append(TimedResult(return_type){
                    .elapsed_total = elapsed_total,
                    .elapsed = elapsed,
                    .value = current_value,
                });
            }

            // add newline for non check-output runs
            if (run_ms > 1) {
                try stderr.writeAll("\n");
            }

            // calculate timings after completing the run
            // (no unnecessary work during benchmark loop)
            const min_ms = Self.calculateMin(&benchmark_results);
            const max_ms = Self.calculateMax(&benchmark_results);
            const mean_ms: f64 = Self.calculateMean(&benchmark_results);
            const std_deviation: f64 = Self.calculateStdDeviation(&benchmark_results, mean_ms);

            return BenchmarkResult(return_type){
                .mean_ms = mean_ms,
                .std_dev_ms = std_deviation,
                .min_ms = min_ms,
                .max_ms = max_ms,
                .runs = benchmark_results.items.len,
                .results = benchmark_results,
            };
        }

        pub fn benchmark(allocator: Allocator, warmup_ms: usize, run_ms: usize, args: anytype) !?BenchmarkResult(return_type) {
            // perform warmup runs
            _ = try Self.run(allocator, warmup_ms, args);

            // perform benchmark runs
            return Self.run(allocator, run_ms, args);
        }

        fn calculateMin(results: *const std.ArrayList(TimedResult(return_type))) f64 {
            var min_ms = std.math.floatMax(f64);

            for (results.items) |result| {
                const milliseconds: f64 = @as(f64, @floatFromInt(result.elapsed)) / 1_000_000.0;

                min_ms = @min(min_ms, milliseconds);
            }

            return min_ms;
        }

        fn calculateMax(results: *const std.ArrayList(TimedResult(return_type))) f64 {
            var max_ms = std.math.floatMin(f64);

            for (results.items) |result| {
                const milliseconds: f64 = @as(f64, @floatFromInt(result.elapsed)) / 1_000_000.0;

                max_ms = @max(max_ms, milliseconds);
            }

            return max_ms;
        }

        fn calculateMean(results: *const std.ArrayList(TimedResult(return_type))) f64 {
            var sum: f64 = 0.0;

            for (results.*.items) |*result| {
                sum += @as(f64, @floatFromInt(result.*.elapsed)) / 1_000_000.0;
            }

            return sum / @as(f64, @floatFromInt(results.*.items.len));
        }

        fn calculateStdDeviation(results: *const std.ArrayList(TimedResult(return_type)), mean_ms: f64) f64 {
            var sum_squares: f64 = 0.0;

            for (results.*.items) |*result| {
                const diff: f64 = (@as(f64, @floatFromInt(result.*.elapsed)) / 1_000_000.0) - mean_ms;
                sum_squares += diff * diff;
            }

            return std.math.sqrt(sum_squares / @as(f64, @floatFromInt(results.*.items.len)));
        }
    };
}

// Helper Functions

pub fn loadArguments(allocator: Allocator) !ArgumentList {
    const args_cli = try std.process.argsAlloc(allocator);

    if (args_cli.len < 4) {
        const stderr = std.io.getStdErr().writer();
        try stderr.print("Usage: {s} <run_ms> <warmup_ms> ...<program args>\n", .{args_cli[0]});
        std.process.exit(1);
    }

    const args = args_cli[1..];

    const run_ms: usize = try std.fmt.parseInt(usize, args[0], 0);
    const warmup_ms: usize = try std.fmt.parseInt(usize, args[1], 0);

    return ArgumentList{
        .run_ms = run_ms,
        .warmup_ms = warmup_ms,
        .program_args = args[2..],
    };
}

pub fn fileReadLines(allocator: Allocator, file_path: []const u8) !std.ArrayList([]const u8) {
    var file = try std.fs.cwd().openFile(file_path, .{ .mode = std.fs.File.OpenMode.read_only });
    defer file.close();

    const stat = try file.stat();
    const file_buffer = try file.readToEndAlloc(allocator, stat.size);

    var word_list = std.ArrayList([]const u8).init(allocator);

    // Split by "\r?\n" and iterate through the resulting slices of "[]const u8"
    var lines = std.mem.tokenizeAny(u8, file_buffer, "\r\n");

    while (lines.next()) |line| {
        try word_list.append(line);
    }

    return word_list;
}
